Mathlib/Data/Finsupp/Weight.lean

```diff
@@ -41,68 +43,83 @@ there are finitely many `f : σ →₀ ℕ` of bounded weight.
 
 ## Degree
 
-- `Finsupp.degree`:  the weight when all components of `w` are equal to `1 : ℕ`.
-The present choice is to have it defined as a plain function.
+- `Finsupp.degree f` is the sum of all `f s`, for `s ∈ f.support`.
+  The present choice is to have it defined as a plain function.
 
 - `Finsupp.degree_eq_zero_iff` says that `f.degree = 0` iff `f = 0`.
 
 - `Finsupp.le_degree` says that `f s ≤ f.degree`.
 
-- `Finsupp.degree_eq_weight_one` says `f.degree = f.weight 1`.
+- `Finsupp.degree_eq_weight_one` says `f.degree = f.weight 1` when `R` is a semiring.
 This is useful to access the additivity properties of `Finsupp.degree`
 
 - For `Finite σ`, `Finsupp.finite_of_degree_le` proves that
 there are finitely many `f : σ →₀ ℕ` of bounded degree.
 
 
-
 ## TODO
 
-* The relevant generality of these constructions is not clear.
-It could probably be generalized to `w : σ → M` and `f : σ →₀ N`,
-provided `N` is a commutative semiring and `M`is an `N`-module.
-
 * Maybe `Finsupp.weight w` and `Finsupp.degree` should have similar types,
 both `AddCommMonoidHom` or both functions.
 
 -/
 
-variable {σ M : Type*} (w : σ → M)
+variable {σ M R : Type*} [Semiring R] (w : σ → M)
 
 namespace Finsupp
 
 section AddCommMonoid
 
-variable [AddCommMonoid M]
-/-- The `weight` of the finitely supported function `f : σ →₀ ℕ`
-with respect to `w : σ → M` is the sum `∑ i, f i • w i`. -/
-noncomputable def weight : (σ →₀ ℕ) →+ M :=
-  (Finsupp.linearCombination ℕ w).toAddMonoidHom
+variable [AddCommMonoid M] [Module R M]
+/-- The `weight` of the finitely supported function `f : σ →₀ R`
+with respect to `w : σ → M` is the sum `∑(f i)•(w i)`. -/
```

Comment:
Can you either change this back to the valid lean code, or write it as a `$ $` \LaTeX block instead?